### ISSUE TITLE
kubectl --validate flag doesn't complain about unknown fields

### ISSUE THREAD

**Issue by a-robinson** [GoogleCloudPlatform/kubernetes#7598]
My understanding of kubectl's --validate flag is that it's supposed to error out if it can't parse its input as a valid API object. Maybe my understanding is wrong, but if it isn't, then the --validate flag is currently broken.

This came up when a user's service was coming up without a load balancer when they used the following yaml:
```
kind: Service
apiVersion: v1beta3
metadata:
  name: locust-master
  labels:
    name: locust
    role: master
spec:
  ports:
    - port: 8089
      targetPort: 8089
      protocol: TCP
      name: locust-master-web
  selector:
    name: locust
    role: master
createExternalLoadBalancer: true
```

The createExternalLoadBalancer parameter is incorrectly at the top level, when it's supposed to be inside the spec. Running `kubectl create -f locust-master-service.yaml --validate` happily succeeds, with the server never seeing the createExternalLoadBalancer parameter.

You can make this even more extreme by throwing in arbitrarily named fields that don't exist anywhere in the API:
```
kind: Service
apiVersion: v1beta3
metadata:
  xxxxxxxxx: yyyyyyyy
  name: locust-master
  labels:
    name: locust
    role: master
spec:
  aaaaaaaa: bbbbbbbbb
  ports:
    - port: 8089
      targetPort: 8089
      protocol: TCP
      name: locust-master-web
  selector:
    name: locust
    role: master
```

Running `kubectl create -f locust-master-service.yaml --validate` with this input file also happily succeeds.

@brendandburns @jlowdermilk

**Comment by derekwaynecarr** [comment]
I ran into a similar issue when using a secret in a pod and had a "source" section in my YAML that was incorrect, but the command did not error, just ignored parts of below irrelevant YAML.

**Comment by nikhiljindal** [comment]
Yes. I have ran into this as well. A field was at the wrong level and --validate just ignored it.
Validation ignores unknown fields: https://github.com/GoogleCloudPlatform/kubernetes/issues/5889

**Comment by a-robinson** [comment]
Isn't #5889 orthogonal? It's about adding server-side validation, whereas I'd like to make our existing client-side validation option validate properly.

**Comment by a-robinson** [comment]
This is also broken for json files, not just yaml:
```
{
   "kind":"Service",
   "apiVersion":"v1beta3",
   "metadata":{
      "name":"locust-master"
   },
   "spec":{
      "ports": [
        {
          "port":8089,
          "targetPort":8089,
          "protocol":"TCP",
          "name": "locust-master-web"
        }
      ],
      "selector":{
         "name":"locust",
         "role":"master"
      }
   },
  "aaaaaaaa":"bbbbbbbb"
}
```

**Comment by nikhiljindal** [comment]
@a-robinson Yes sorry. You are right.

**Comment by bgrant0607** [comment]
I believe the current validation at most catches required fields that are not provided. Unfortunately, our nearly ubiquitous use of omitempty defeats even that checking.

**Comment by bgrant0607** [comment]
@brendandburns Are you actively working on this? If not, we'd like to reassign it.

Yesterday, we discussed stricter json parsing (also discussed in #3338) as an alternative. Since kubectl currently round-trips objects (#3955), it will drop unknown fields, anyway.

However, when I was thinking about #4855, it reminded me that strict parsing needs to be optional, and we'd want to disable it at least for parsing responses in essentially all clients. We might also want to make it optional for requests (#5889), but I haven't thought of a use case for that yet.

**Comment by a-robinson** [comment]
I asked about whether strict parsing should be the default for requests in #7600, some concerns were brought up there.

**Comment by bgrant0607** [comment]
Also being discussed in #4855.

**Comment by brendandburns** [comment]
@bgrant0607 I believe that this specific issue is fixed by #7919 I'm going to close it, we can discuss the defaulting elsewhere.

fwiw, making it optional for parsing requests is useful too in case you have skew between client version and server version.

Also, we should likely have a send the raw data we read, rather than re-encoding the object for transmittal, option.
